packages/lsp/src/tokens.ts

```diff
@@ -356,6 +357,11 @@ export interface IDocumentConnectionData {
  * The LSP connection state manager
  */
 export interface ILSPDocumentConnectionManager {
+  /**
+   * The path of the active widget adapter.
+   */
+  currentAdapter: string | null;
```

Comment:
I left three comments explaining it above. 

> The current adapter contains the path to the active adapter instead of the adapter itself, because when adding a new adapter, `app.shell.currentChanged` is emitted before the adapter is registered.
> *Posted [here](https://github.com/jupyterlab/jupyterlab/pull/14920#discussion_r1284318643)*

Reply:
Sorry I missed it. IMO this shows that the logic needs to be rewritten. For such edge cases instead of emitting when `shell.currentChanged` is fired, we should emit when adapter is first registered if it is current widget of the shell. This is I believe it should be responsibility of the tracker to emit signal when adapter actually changed, not when it "could have changed if it was already registered".